# CIHEAM2015
Code and tutorials for CIHEAM course. QC, Variation calling and filtering sessions

# Index

QC of NGS reads

Exercise: Run FastQC to check your data

Exercise: Run Trimmomatic to filter your data

Introduction to Variation Calling

Exercise: Prepare a BAM file only with the reads mapped on chromosome 18

Exercise: Run FreeBayes

Exercise: Run GATK UnifiedGenotyper

Exercise: Run Samtools Variation Calling

VCF Filtering

Exercise: Select only SNP from a VCF file

Exercise: Compress and index a VCF file

Exercise: Filter the VCF file according to different parameters

Exercise: Run vcf-compare to get statistics on different VCF files

Exercise: Plot the results of vcf-compare using R



QC of NGS reads
===============

Exercise: Run FastQC to check your data
---------------------------------------

Exercise: Run Trimmomatic to filter your data
---------------------------------------------

Introduction to Variation Calling
=================================

Exercise: Prepare a BAM file only with the reads mapped on chromosome 18
------------------------------------------------------------------------

Exercise: Run FreeBayes
-----------------------

Exercise: Run GATK UnifiedGenotyper
-----------------------------------

Exercise: Run Samtools Variation Calling
----------------------------------------

VCF Filtering
=============

Exercise: Select only SNP from a VCF file
-----------------------------------------

Exercise: Compress and index a VCF file
---------------------------------------

Exercise: Filter the VCF file according to different parameters
---------------------------------------------------------------

Exercise: Run vcf-compare to get statistics on different VCF files
------------------------------------------------------------------

Exercise: Plot the results of vcf-compare using R
-------------------------------------------------
